### PR TITLE
Fix deprecation warning on Python 3.11

### LINKED ISSUE #192: DeprecationWarning from importlib.resources on Python 3.11
```
$ python3.11 -W all -c 'import certifi ; certifi.where()'
/.../python3.11/site-packages/certifi/core.py:36: DeprecationWarning: path is deprecated. Use files() instead. Refer to https://importlib-resources.readthedocs.io/en/latest/using.html#migrating-from-legacy for migration advice.
  _CACERT_CTX = get_path("certifi", "cacert.pem")
```

### PR DESCRIPTION
**Pull request by adamchainz**
Fixes #192.

### CODE REVIEW

**Comment by adamchainz** [comment]
I ran tests on Python 3.7+ locally, checking for warnings, with:

```
[tox]
envlist =
    py{37,38,39,310,311}

[testenv]
commands = pytest
deps = pytest
setenv =
    PYTHONDEVMODE=1
```

None emitted.

**Comment by adamchainz** [comment]
Probably #167 should be merged first.

**Comment by hugovk** [comment]
@adamchainz https://github.com/certifi/python-certifi/pull/167 has been merged.

**Comment by adamchainz** [comment]
Rebased!

**Review by AA-Turner — APPROVED** [review]
3.6 is now EOL, so you could just drop the final fallback & support for 3.6 -- every upstream supported version has a version of `importlib.resources`.

A

**Comment by sigmavirus24** [comment]
> 3.6 is now EOL, so you could just drop the final fallback & support for 3.6 -- every upstream supported version has a version of `importlib.resources`.
> 
> A

Not every supported version has the same API in the standard library because it's API is explicitly not stable until maybe 3.10.  The fallback is necessary unless we drop everything<3.10 and even then there's a commitment to moving the API quickly to something "better" if something seemingly better is identified without concern for stability promises

**Comment by AA-Turner** [comment]
Sorry, by the final fallback I meant the `__file__` version -- the `path` API will remain stable in Python 3.7 and 3.8, I believe?

A

**Comment by sigmavirus24** [comment]
> Sorry, by the final fallback I meant the `__file__` version -- the `path` API will remain stable in Python 3.7 and 3.8, I believe?
> 
> A

It changed after that, and may change in 3.12, 3.13 or some other future version

File: certifi/core.py [hunk 1]
```diff
@@ -1,15 +1,18 @@
+# -*- coding: utf-8 -*-
+
```
[the remark this answers is not in the document]
Comment (adamchainz) [reply certifi/core.py:2]: thanks for spotting, I think it got added back accidentally in rebasing

**Comment by adamchainz** [comment]
Okay, I've dropped Python 3.6 support, added the 3.11 tag, and removed the Python <3.7 branch.

File: setup.py [hunk 2]
```diff
@@ -46,19 +46,19 @@
...
     license='MPL-2.0',
-    python_requires=">=3.6",
+    python_requires=">=3.7",
     classifiers=[
         'Development Status :: 5 - Production/Stable',
         'Intended Audience :: Developers',
         'License :: OSI Approved :: Mozilla Public License 2.0 (MPL 2.0)',
         'Natural Language :: English',
         'Programming Language :: Python',
         'Programming Language :: Python :: 3',
```
Comment (AA-Turner) [inline setup.py:56]:
```suggestion
        'Programming Language :: Python :: 3',
        'Programming Language :: Python :: 3 :: Only',
```

A

File: certifi/core.py [hunk 3]
```diff
@@ -33,36 +33,44 @@ def where() -> str:
             # We also have to hold onto the actual context manager, because
             # it will do the cleanup whenever it gets garbage collected, so
             # we will also store that at the global level as well.
-            _CACERT_CTX = get_path("certifi", "cacert.pem")
+            _CACERT_CTX = as_file(files("certifi").joinpath("cacert.pem"))
             _CACERT_PATH = str(_CACERT_CTX.__enter__())
 
         return _CACERT_PATH
```
Comment (AA-Turner) [inline certifi/core.py:40]:
```suggestion


    def contents() -> str:
        return files("certifi").joinpath("cacert.pem").read_bytes()

```

File: certifi/core.py [hunk 4]
```diff
@@ -33,36 +33,44 @@ def where() -> str:
...
+            # __exit__() is a no-op.
+            #
+            # We also have to hold onto the actual context manager, because
+            # it will do the cleanup whenever it gets garbage collected, so
+            # we will also store that at the global level as well.
+            _CACERT_CTX = get_path("certifi", "cacert.pem")
+            _CACERT_PATH = str(_CACERT_CTX.__enter__())
 
-        return os.path.join(f, "cacert.pem")
+        return _CACERT_PATH
```
Comment (AA-Turner) [inline certifi/core.py:72]:
```suggestion


    def contents() -> str:
        read_binary("certifi", "cacert.pem")

```

File: certifi/core.py [hunk 5]
```diff
@@ -33,36 +33,44 @@ def where() -> str:
...
             _CACERT_PATH = str(_CACERT_CTX.__enter__())
 
         return _CACERT_PATH
 
+else:
 
-except ImportError:
-    Package = Union[types.ModuleType, str]
-    Resource = Union[str, "os.PathLike"]
+    from importlib.resources import path as get_path
```
Comment (AA-Turner) [inline certifi/core.py:43]:
```suggestion
    from importlib.resources import path as get_path, read_binary
```

**Review by AA-Turner — COMMENTED** [review]
This suggests inlining `contents()` into the version checks using the native `importlib.resources` APIs, rather than coercing to a filesystem path.

I can't put a GH suggestion to delete the top level `contents()` as it covers added and deleted lines.

A

File: setup.py (hunk 2 above)
Comment (adamchainz) [reply setup.py:56]: sure but I'm not sure anyone cares about that classifier any more

File: certifi/core.py (hunk 4 above)
Comment (adamchainz) [reply certifi/core.py:72]:
```suggestion


    def contents() -> str:
        return read_text("certifi", "cacert.pem", encoding="ascii")
```

File: setup.py (hunk 2 above)
Comment (AA-Turner) [reply setup.py:56]: Fair enough, I'm never really sure with classifiers.

A

**Comment by adamchainz** [comment]
Squashed everything together, ready for merge.

**Review by AA-Turner — APPROVED** [review]

**Comment by alex** [comment]
Unfortunately python 3.6 still represents 15% of our downloads, so I think I speak for all the maintainers when I say we're going to be hesitant about the tradeoffs of fixing deprecation warnings by dropping support for such a large group of users.

**Comment by AA-Turner** [comment]
Would the PR be accepted if it incorporated the filesystem branch?

A

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -22,8 +22,12 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: ["3.6", "3.7", "3.8", "3.9", "3.10"]
-
+        python-version:
+        - "3.7"
```
Comment (AA-Turner) [inline .github/workflows/ci.yml:26]:
```suggestion
        - "3.6"
        - "3.7"
```

File: setup.py [hunk 7]
```diff
@@ -46,19 +46,20 @@
     include_package_data=True,
     zip_safe=False,
     license='MPL-2.0',
-    python_requires=">=3.6",
+    python_requires=">=3.7",
```
Comment (AA-Turner) [inline setup.py:49]:
```suggestion
    python_requires=">=3.6",
```

File: certifi/core.py [hunk 8]
```diff
@@ -33,36 +33,45 @@ def where() -> str:
...
-            _CACERT_CTX = get_path("certifi", "cacert.pem")
+            _CACERT_CTX = as_file(files("certifi").joinpath("cacert.pem"))
             _CACERT_PATH = str(_CACERT_CTX.__enter__())
 
         return _CACERT_PATH
 
+    def contents() -> str:
+        return files("certifi").joinpath("cacert.pem").read_text(encoding="ascii")
+
+else:
```
Comment (AA-Turner) [inline certifi/core.py:44]:
```suggestion
elif sys.version_info >= (3, 7):
```

**Review by AA-Turner — COMMENTED** [review]
Suggested changes restore 3.6 compat. (untested)

A

**Review by ssbarnea — APPROVED** [review]

**Comment by ssbarnea** [comment]
Can we please proceed with this? This seems to be a blocker for enabling us to test against py311 as we treat warnings as errors.

**Comment by alex** [comment]
This PR breaks Python 3.6 support as-is. We have said several times that it will not be considered until 3.6 support is restored.

**Comment by ssbarnea** [comment]
@alex py36 reached end-of-life [more than 6 months ago](https://endoflife.date/python), why not just dropping it? If someone wants to maintain it, it can be done in a maintenance branch.

IMHO, we should to hinder main (brand) development on a dead snake, especially as it can be quite tricky to keep compatibility across a wide range of python versions.

The way I see it is that original PR is more than two months old and soon we might end-up not having a certify for python 3.11 if we don't fix it. What is worse? Still, if agreed, removal of support for py36 should be done in a separated change.

**Comment by hugovk** [comment]
Please see PR https://github.com/certifi/python-certifi/pull/199, which is the same as this with @AA-Turner's suggestions applied to retain 3.6 support.

**Comment by hugovk** [comment]
Here's the download share for different versions, 3.6 is the third biggest chunk:

image

https://hugovk.github.io/pypi-tools/charts.html#certifi

**Comment by ssbarnea** [comment]
Thanks for addressing it. Regarding downloads stats, please be aware that they might be painting a picture very different than what people writing code might attempt to use.

- if you do not have wheels for a version, it will not be listed, even if people might attempt to use it
- if a specific version is buggy or even broken it will likely not get so many downloads
- pypi traffic comes from >95% CI/CD implementations for most libraries (exception those "end-user" packages that are not really re-used). I ended up being listed with 16 critical projects, even if some of these were even projects with 1-2 contributors and less than 10 stars on github, basically being spin-offs from our main projects. The only reason for that is because they got a huge number of downloads from our CI/CD pipelines integration (they are also testing tools, so others reuse them for testing).
- what the graph is telling you is who is making traffic, many of the old versions are from CI/CD pipelines still using them, often even scheduled ones running on dead projects (yep, seen that -- at least github introduced an option to disable actions after 6mo of inactivity)

That is why I suggested moving an old version to a maintenance branch, so if someone really needs a fix, they would have to backport it, likely having to make other changes to do so, maybe even sponsor/pay maintainers to get it done. I am a maintainer myself and I consider myself (morally) responsable to help others that contribute back. When it comes to support "old" stuff, that is where you can be picky. I seen some very small projects keeping compatibility with only 2 versions of python but that is extreme. Obviously that "certify" is an AAA class project, at least as popularity, not budget, so doing that would be insane.

Again, a bit thanks to all involved!

**Comment by alex** [comment]
It is incorrect that not having wheels impacts download stats. PyPI download stats are computed using the Python version doing the install, from its user agent, not based on the Python version a wheel targets. Indeed, certifi only publishes universal wheels.

Further, neither Python 3.9 nor 3.10 are buggy, so we wouldn't expect that to depress download counts.

While its true that many downloads come from CI systems, these are nonetheless systems that would be impacted by dropping a Python version.

Finally, a maintenance branch makes little sense for this project, given that our primary purpose is to distribute a PEM file. It's not even clear to me how we'd version such a branch.

**Comment by ssbarnea** [comment]
I totally agree with you. Probably this project is the best use-case against using a maintenance branch, as it basically packages a single data file, nothing else.

**Comment by ssbarnea** [comment]
Any chance of doing a patch release so we can start using with py311? I would prefer to not have to add an ignore for the fixed runtime warning.

**Pull request closed without merging**